src/Common/HashTable/HashTable.h

```diff
@@ -319,16 +320,20 @@ class alignas(64) HashTableGrowerWithPrecalculation
     bool overflow(size_t elems) const { return elems > precalculated_max_fill; }
 
     /// Increase the size of the hash table.
-    void increaseSize() { increaseSizeDegree(size_degree >= 23 ? 1 : 2); }
+    void increaseSize() { increaseSizeDegree(size_degree >= max_size_degree_quadratic ? 1 : 2); }
 
     /// Set the buffer size by the number of elements in the hash table. Used when deserializing a hash table.
     void set(size_t num_elems)
     {
-        size_degree = num_elems <= 1
-             ? initial_size_degree
-             : ((initial_size_degree > static_cast<size_t>(log2(num_elems - 1)) + 2)
-                 ? initial_size_degree
-                 : (static_cast<size_t>(log2(num_elems - 1)) + 2));
+        if (num_elems <= 1)
+            size_degree = initial_size_degree;
+        else if (initial_size_degree > static_cast<size_t>(log2(num_elems - 1)) + 2)
+            size_degree = initial_size_degree;
+        else
+        {
+            size_degree = static_cast<size_t>(log2(num_elems - 1));
```

Comment:
if `num_elems = 2^24`, `size_degree` will be `log2(2^24 -1) + 1 = 24`, so there will be no extra space in the HT and a good amount of collisions.

Reply:
@nickitat nice catch! How about the following?

```python
int(math.ceil(math.log2(n) + 0.5)))
```

This will give the following:

```
size:           256 (2^   8.0), upstream:  9 (load_factor:  0.5), patched:  9 (load_factor:  0.5)  
size:   2,516,582.4 (2^ 21.26), upstream: 23 (load_factor:  0.3), patched: 22 (load_factor:  0.6) CHANGED
size:   2,936,012.8 (2^ 21.49), upstream: 23 (load_factor: 0.35), patched: 22 (load_factor:  0.7) CHANGED
size:   3,355,443.2 (2^ 21.68), upstream: 23 (load_factor:  0.4), patched: 23 (load_factor:  0.4)  
size:   3,774,873.6 (2^ 21.85), upstream: 23 (load_factor: 0.45), patched: 23 (load_factor: 0.45)  
size:     4,194,304 (2^  22.0), upstream: 23 (load_factor:  0.5), patched: 23 (load_factor:  0.5)  
size:   5,033,164.8 (2^ 22.26), upstream: 24 (load_factor:  0.3), patched: 23 (load_factor:  0.6) CHANGED
size:   7,549,747.2 (2^ 22.85), upstream: 24 (load_factor: 0.45), patched: 24 (load_factor: 0.45)  
size:     8,388,608 (2^  23.0), upstream: 24 (load_factor:  0.5), patched: 24 (load_factor:  0.5)  
size:  10,066,329.6 (2^ 23.26), upstream: 25 (load_factor:  0.3), patched: 24 (load_factor:  0.6) CHANGED
size:  15,099,494.4 (2^ 23.85), upstream: 25 (load_factor: 0.45), patched: 25 (load_factor: 0.45)  
size:    16,777,216 (2^  24.0), upstream: 25 (load_factor:  0.5), patched: 25 (load_factor:  0.5)  
```

It slightly increases the load factor when the size closer to the previous degree

<details>

<summary>test-ch-grower.py</summary>

```python
#!/usr/bin/env python3

import math

def grower_upstream_degree(n):
    return int(math.log2(n - 1)) + 2

def grower_patched_degree(n):
    return int(math.ceil(math.log2(n) + 0.5))

if __name__ == '__main__':
    for size in [
        (1<<8),
        (1<<22) * 0.6,
        (1<<22) * 0.7,
        (1<<22) * 0.8,
        (1<<22) * 0.9,
        (1<<22),
        (1<<23) * 0.6,
        (1<<23) * 0.9,
        (1<<23),
        (1<<24) * 0.6,
        (1<<24) * 0.9,
        (1<<24),
    ]:
        upstream = grower_upstream_degree(size)
        patched = grower_patched_degree(size)
        assert size >= upstream
        assert size >= patched
        print('size: {:13,} (2^{:6.4}), upstream: {:2} (load_factor: {:4.2}), patched: {:2} (load_factor: {:4.2}) {}'.format(
            size, math.log2(size),
            upstream, size/(1<<upstream),
            patched, size/(1<<patched),
            'CHANGED' if upstream != patched else ' '))
```

</details>

Reply:
for performance I think having more than 0.5 fill factor is a bad idea. do you really see much benefit from this change? in the end of the day each HT cell is just a key and a pointer to an aggr func state, so we're talking about couple of GBs per 1e8 elements if I did the math correctly.

Reply:
>for performance I think having more than 0.5 fill factor is a bad idea

Everyone has it's own constant :)

But just out of curiosity I decided to check this, and indeed performance wise it is better to use smaller load factor, with 0.3 load factor HashMap<UInt64, UInt16> with 10e6 elements works 50% slower then HashMap with 0.6 load factor.

<details>

Note, that this setting for dictionaries is not available yet.

```sql
CREATE TABLE default.dict_data ( `key` UInt64, `value` UInt16 ) ENGINE = GenerateRandom;

create dictionary hash_map_grow_0_7 (key UInt64, value UInt16) primary key key SOURCE(CLICKHOUSE(TABLE dict_data WHERE '1 limit 10e6')) LIFETIME(MIN 0 MAX 0) LAYOUT(HASHED(MAX_LOAD_FACTOR 0.7)) SETTINGS(max_threads = 1);
create dictionary hash_map_grow_0_5 (key UInt64, value UInt16) primary key key SOURCE(CLICKHOUSE(TABLE dict_data WHERE '1 limit 10e6')) LIFETIME(MIN 0 MAX 0) LAYOUT(HASHED(MAX_LOAD_FACTOR 0.5)) SETTINGS(max_threads = 1);

select count() from (select key, dictGet(hash_map_grow_0_7, 'value', key) dict_value, value from (select * from dict_data where 1 limit 10e6)) where value != dict_value settings max_threads=1;
-- 1 row in set. Elapsed: 0.378 sec. Processed 10.01 million rows, 100.08 MB (26.45 million rows/s., 264.48 MB/s.)

select count() from (select key, dictGet(hash_map_grow_0_5, 'value', key) dict_value, value from (select * from dict_data where 1 limit 10e6)) where value != dict_value settings max_threads=1;
-- 1 row in set. Elapsed: 0.255 sec. Processed 10.01 million rows, 100.08 MB (39.19 million rows/s., 391.92 MB/s.)
```

</details>

>do you really see much benefit from this change?

In terms of memory - yes, but, it needed this to allow configuring `max_load_factor` for `hashed`/`sparse_hashed` dictionaries (#49380), but it is not required anymore, since I wrote there custom grower (anyway it is required to change the overflow condition).

> in the end of the day each HT cell is just a key and a pointer to an aggr func state, so we're talking about couple of GBs per 1e8 elements if I did the math correctly

You are talking about Aggregator here, well, few GiBs per query can be significant.

But I will close it for now, since I solved my initial problem with a custom grower anyway.